chore: add devenv.sh development environment

landure opened the pull request:
add [devenv.sh](https://devenv.sh/) Nix-based development environment with:

- PHP 8.1 with the least possible enabled extensions.
- PostgreSQL server (not in a Docker container).
- [pre-commit](https://pre-commit.com/) commit hooks for PHP-CS-Fixer, PHPStan, and Rector.

This merge request aims to provide a standardized PHP 8.1 development environment for contributors.
It can help creating a dev container for this project.

landure commented on devenv.nix:
```diff
@@ -0,0 +1,185 @@
+  };
+
+  # https://devenv.sh/packages/
+  packages = with pkgs; [ git ];
+
+  # https://devenv.sh/languages/
```
devenv.sh language.php.enable install composer with php.

martin-georgiev commented:
Thank you @landure for this PR. It certainly moves the project in a more contributor-friendly direction. 

The AI code reviewer dropped some suggestions. AI being AI can be overly verbose and nitpicking sometimes. What do you think of the latest suggestions CodeRabbit had added?

landure commented on docs/CONTRIBUTING.md:
```diff
@@ -1,39 +1,160 @@
+
+### Local development
+
+ℹ️ Use `devenv.local.nix` to alter the development environment.
+For example, this file:
+
+- Install [Harlequin](https://harlequin.sh/) database TUI.
+- Set PHP version to 8.4.
+- Change PostgreSQL related environment variables.
+
```
direnv.local.nix is present in the gitignore, and can store plaintext passwords for local development.

landure commented:
Cachix should be enable by default, but it requires adding a trusted key to nix package manager configuration.
I've noted the needed commands as being the following, but at the moment, I've got no way to test this:

Trust [Cachix DevEnv](https://devenv.sh/) store:

```bash
sudo tee -a '/etc/nix/nix.conf' <<EOF
# Trust Cachix DevEnv
extra-substituters = https://devenv.cachix.org
extra-trusted-public-keys = devenv.cachix.org-1:w1cLUi8dv3hnoSPGAuibQv+f9TZLr6cv/Hm9XgU50cw=

EOF
```

Restart `nix-daemon` to load the new configuration:

```bash
sudo systemctl 'restart' 'nix-daemon.service'
```

landure commented on docs/CONTRIBUTING.md:
````diff
@@ -1,39 +1,160 @@
 
+   ```bash
+   nix-env --install --attr devenv -f https://github.com/NixOS/nixpkgs/tarball/nixpkgs-unstable
+   ```
````
I've kept nix-env method, since nix command is considered experimental. I've added nix profile command as given in devenv.sh setup page as being the recommended method, and update the nix install guide with the required settings to enable nix command.

landure commented:
@martin-georgiev being quite nitpicking myself when in the appropiate mood, I think the current state of the PR is quite OK. I've added nix package manager recommended configuration, with cachix trusted key. 
I've also added commitizen and composer validate and audit git hooks.
I've added sample mdformat and markdownlint hooks, but not enabled them.

martin-georgiev commented on docs/CONTRIBUTING.md:
```diff
@@ -1,39 +1,233 @@
+## 🧑‍💻 Development with Devenv
 
-For the sake of clear Git history and speedy review of your PR, please check that the suggested changes are in line with the project's standards. Code style, static analysis and file validation scripts are already provided and can easily be run from project's root:
+This project supports [devenv.sh](https://devenv.sh/) for a consistent
+development environment:
```
```suggestion
This project supports [devenv.sh](https://devenv.sh/) for a consistent development environment:
```

martin-georgiev commented on docs/CONTRIBUTING.md:
```diff
@@ -1,39 +1,233 @@
+development environment:
 
-`composer check-code-style` which will check for consistent code style
+1. Install the [Nix package manager](https://nixos.org/download/#download-nix)
+   (if not already installed).
```
### 
```suggestion
1. Install the [Nix package manager](https://nixos.org/download/#download-nix) (if not already installed).
```

martin-georgiev commented on docs/CONTRIBUTING.md:
````diff
@@ -1,39 +1,233 @@
+        sudo launchctl kickstart -k system/org.nixos.nix-daemon
+        ```
+
+2. Install [devenv.sh](https://devenv.sh/) (if not already installed),
+   by following [Getting started @ devenv.sh](https://devenv.sh/getting-started/):
````
```suggestion
2. Install [devenv.sh](https://devenv.sh/) (if not already installed), by following [Getting started @ devenv.sh](https://devenv.sh/getting-started/):
```

martin-georgiev commented on docs/CONTRIBUTING.md:
```diff
@@ -1,39 +1,233 @@
-3. Depending on the new data-type nature you may have to overwrite some of the following methods:
+2. Give the new data type a unique name within your application.
+   Use the `TYPE_NAME` constant for that purpose.
+3. Depending on the new data-type nature you may have to overwrite some of
+   the following methods:
```
```suggestion
3. Depending on the new data-type nature you may have to overwrite some of the following methods:
```

martin-georgiev commented on docs/CONTRIBUTING.md:
```diff
@@ -1,39 +1,233 @@
 
-----
-**How to add more functions?**
+Most new functions will likely have a signature very similar to those already
+implemented in the project.
```
```suggestion
Most new functions will likely have a signature very similar to those already implemented in the project.
```

martin-georgiev commented on docs/CONTRIBUTING.md:
````diff
@@ -49,11 +243,13 @@ class ArrayAppend extends BaseFunction
 ```
 
-*Beware that you cannot use **?** (e.g. the `??` operator) as part of any function prototype in Doctrine as this will fail the query parsing.*
+⚠️ **Beware:** you cannot use **?** (e.g. the `??` operator) as part of any
+function prototype in Doctrine.
````
```suggestion
⚠️ **Beware:** you cannot use **?** (e.g. the `??` operator) as part of any function prototype in Doctrine.
```
